### DISCUSSION TITLE
How to set 32 bit per sample?

### DISCUSSION THREAD

**Discussion by constant-flow** [pschatzmann/arduino-snapclient#18] [Q&A] [answered]
Hi, the DAC I use (TAS5713) requires a high clock rate for the MCLK as i learned from the squeezelite project.

> The tas5713 typically has the mclk connected to the sclk. In this
> configuration, mclk must be a multiple of the sclk. The lowest workable
> multiple is 64x. To achieve this,  32 bits per channel on must be sent
> over I2S. Reconfigure the I2S for that here, and expand the I2S stream
> when it is sent

Doing this based of from the sample code provided here

```
auto cfg = out.defaultConfig();
...
cfg.bits_per_sample = 32;
```

Is not sufficient to convince the lib to start 32 bit sample stream, it always falls back to 16.
How to setup the stream properly with something like the expansion as mentioned above?

**Comment by pschatzmann** [comment, marked as answer]
The output is configured by the information by the server, so if you set it to 32 bits it will be overwritten.

If you want resample on the client side to change the format from 16 to 32 bits, Just wrap the output class (e.g. I2SStream) with a FromatConverterStream or better a NumberFormatConverterStream and  provide the converterstream as output class to the snap client.  Don't forget to configure the converter class by defining the input and output format and the output format of i2s

Further info can be fund in the [Wiki of the AudioTools](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format).

Another option would be to change the format in the snapserver.conf. Then you will not have to change anything on the client side.

**Comment by pschatzmann** [comment]
I was testing today the resampling from 16 to 32 for a ES9018k2m DAC which has the same problems and I had to commit a couple of corrections to the audio-tools to make this work.

Here is [the example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-url_mp3_helix-i2s_32bit/streams-url_mp3_helix-i2s_32bit.ino) that I used for testing

**Reply by constant-flow** [reply]
Thanks for the hints/links, the testing and corrections. Much appreciated, I'll try to test the code as soon as possible, maybe this night.

**Reply by constant-flow** [reply]
Hi I tried to compile the provided test code. Line 22 gives me errors, as it expects a `Print` object as it seems. 

```
StreamCopy copier(dec, url); // copy url to decoder
                           ^
In file included from .pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools.h:83,
                 from src/main.cpp:2:
.pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools/StreamCopy.h:408:9: note: candidate: 'audio_tools::StreamCopy::StreamCopy(Print&, Stream&, int)'
         StreamCopy(Print &to, Stream &from, int buffer_size=DEFAULT_BUFFER_SIZE) : StreamCopyT<uint8_t>(to, from, buffer_size){
         ^~~~~~~~~~
.pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools/StreamCopy.h:408:9: note:   no known conversion for argument 2 from 'audio_tools::URLStream()' to 'Stream&'
.pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools/StreamCopy.h:404:9: note: candidate: 'audio_tools::StreamCopy::StreamCopy(Print&, audio_tools::AudioStream&, int)'
         StreamCopy(Print &to, AudioStream &from, int buffer_size=DEFAULT_BUFFER_SIZE) : StreamCopyT<uint8_t>(to, from, buffer_size){
         ^~~~~~~~~~
.pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools/StreamCopy.h:404:9: note:   no known conversion for argument 2 from 'audio_tools::URLStream()' to 'audio_tools::AudioStream&'
.pio/libdeps/wt32-eth01/arduino-audio-tools/src/AudioTools/StreamCopy.h:400:9: note: candidate: 'audio_tools::StreamCopy::StreamCopy(int)'
         StreamCopy(int buffer_size=DEFAULT_BUFFER_SIZE): StreamCopyT<uint8_t>(buffer_size) {
```

**Comment by pschatzmann** [comment]
I don't get any errors and your error messages do not make any sense to me: dec extends from Print and url from AudioStream, so the second constructor should be valid! Even the first should work since AudioStream extends from Stream...

There must be something completely wrong with your installation of the AudioTools library

Did you really use the example w/o any changes ?

**Reply by constant-flow** [reply]
As i have it hooked up to my WT32-ETH01 and my DAC which requires the I2C initialization phase I implanted your code into my base (... and I was too lazy to repatch the dupont wires :| )

As I am relying on an Ethernet connection i removed the SSID and passphrase so I remained with `URLStream url();` ... this causes then the error I posted, ... so probably 100% unrelated but maybe another issue for later :sweat_smile: 

So it compiled, but my DAC wasn't putting anything out yet. Will test tomorrow.

**Reply by constant-flow** [reply]
I rewired everything to "normal" ESP32dev board and the provided sketch works flawlessly after the I2C initialization (see `tas5713_init_sequence` in [TAS5713.h](https://github.com/thenaran/linux-rpi/blob/master/sound/soc/codecs/tas5713.h))
